### PR TITLE
BSP-1480, BSP-3125: Adds ability to bulk upload mulitple file types.

### PR DESCRIPTION
**Pull request by rohitmonga**
Dependent on [Dari#291](https://github.com/perfectsense/dari/pull/291)

### CODE REVIEW

File: db/src/main/java/com/psddev/cms/tool/page/content/Upload.java [hunk 1]
```diff
@@ -341,27 +382,113 @@ private static void writeFilePreview(ToolPageContext page) throws IOException, S
...
+
+                if (smartUploadableTypes.stream().noneMatch(t -> t.getType().equals(type) && hasMimeType(t.getField(), fileMimeType))) {
+                    continue;
+                }
+
+                Set<SmartUploadableType> compatibleTypes = smartUploadableTypes.stream()
+                        .filter(t -> hasMimeType(t.getField(), fileMimeType))
+                        .collect(Collectors.toSet());
+
+                // File should be mapped to field with most specific mime type.
```
Comment (rhino88) [inline db/src/main/java/com/psddev/cms/tool/page/content/Upload.java]: Here is another scenario where we probably need to provide a choice, vs. choosing for the user.

File: db/src/main/java/com/psddev/cms/tool/page/content/Upload.java [hunk 2]
```diff
@@ -341,27 +382,113 @@ private static void writeFilePreview(ToolPageContext page) throws IOException, S
...
+            }
+
+            if (smartUploadableTypes != null) {
+                String fileMimeType = file.getContentType();
+
+                if (smartUploadableTypes.stream().map(SmartUploadableType::getField).noneMatch(field -> hasMimeType(field, fileMimeType))) {
+                    throw new IllegalArgumentException("Invalid mime type(s)!");
                 }
+
+                if (smartUploadableTypes.stream().noneMatch(t -> t.getType().equals(type) && hasMimeType(t.getField(), fileMimeType))) {
```
Comment (rhino88) [inline db/src/main/java/com/psddev/cms/tool/page/content/Upload.java]: This logic can be improved to not use so many streams on the same collection

File: db/src/main/java/com/psddev/cms/tool/page/content/Upload.java [hunk 3]
```diff
@@ -341,27 +382,113 @@ private static void writeFilePreview(ToolPageContext page) throws IOException, S
...
+                    continue;
+                }
+
+                Set<SmartUploadableType> compatibleTypes = smartUploadableTypes.stream()
```
Comment (rhino88) [inline db/src/main/java/com/psddev/cms/tool/page/content/Upload.java]: Same comment as above

File: db/src/main/java/com/psddev/cms/tool/page/content/Upload.java [hunk 4]
```diff
@@ -341,27 +382,113 @@ private static void writeFilePreview(ToolPageContext page) throws IOException, S
...
+                        .filter(t -> hasMimeType(t.getField(), fileMimeType))
+                        .collect(Collectors.toSet());
+
+                // File should be mapped to field with most specific mime type.
+                if (compatibleTypes.size() > 1 && compatibleTypes.stream()
```
Comment (rhino88) [inline db/src/main/java/com/psddev/cms/tool/page/content/Upload.java]: Performing another stream on the same collection as above

File: db/src/main/java/com/psddev/cms/tool/page/content/Upload.java [hunk 5]
```diff
@@ -341,27 +382,113 @@ private static void writeFilePreview(ToolPageContext page) throws IOException, S
...
+        return new SparseSet(StringUtils.isBlank(mimeTypes) ? "+/" : mimeTypes).contains(mimeType);
     }
 
     public enum Context {
         FIELD,
         GLOBAL
     }
+
+    @Recordable.Abstract
+    private static class SmartUploadableType extends Record {
```
Comment (rhino88) [inline db/src/main/java/com/psddev/cms/tool/page/content/Upload.java]: Does this need to be a Record?

Comment (rohitmonga) [inline db/src/main/java/com/psddev/cms/tool/page/content/Upload.java]: Yes, otherwise there is an `java.lang.IncompatibleClassChangeError`.

**Comment by rohitmonga** [comment]
Closing in favor of [BSP-47](https://github.com/perfectsense/brightspot/pull/47).

**Pull request closed without merging**